src/services/API/RootClient.js

```diff
@@ -77,34 +81,59 @@ client.interceptors.response.use(
-            var newAccessToken = store.state.auth.accessToken.access_token;
-            // Add the new access token to the header of the request
-            error.config.headers["Authorization"] = `Bearer ${newAccessToken}`;
-
+      // reAuthenticate the user if this is the first reAuthentication call. While reAuthentication is going on
+      // if some other request ends up here, then wait for the reAuthentication process to finish. Once that is done,
+      // set the newly retrieved access token as an auth header and retry the request. Any errors in the reauthentication
+      // process are caught and the user is logged out.
+      let run = async () => {
+        try {
+          let reAuthenticationState = store.state.auth.reAuthenticationState;
+          switch (reAuthenticationState) {
+            case "dropped":
+              // re authentication process was dropped, throw an error.
+              throw new Error(
+                "Re-authentication failed, no more calls allowed"
+              );
+            case "in-process":
+              // re authentication process in progress. Wait for the promise to resolve,
+              // then proceed
+              await store.state.auth.reAuthenticationPromise;
+              break;
+            case "not-started":
+              // fresh re authentication process to begin now. Wait for the process to complete,
+              // then proceed
+              await UserFunctionalService.reAuthenticate(store);
+              break;
+          }
+          // Add the new access token recieved from the re authentication process
+          // to the header of the request and retry the request
+          let newAccessToken = store.state.auth.accessToken.access_token;
+          error.config.headers["Authorization"] = `Bearer ${newAccessToken}`;
+          if (error.config.url == userFromTokenEndpoint) {
             // If the call is fetching a user from an access token, we need to update the
             // request params with the new access token
-            if (error.config.url == userFromTokenEndpoint)
-              error.config.params["token"] = newAccessToken;
-
-            // try the request again
-            return client.request(error.config);
+            error.config.params["token"] = newAccessToken;
+          }
+          return client.request(error.config); // retrying the request
+        } catch (err) {
+          // an error occured in the re authentication process
+          // reset the reAuthenticationState value, so any future reAuthentication processes
+          // can function normally
+          store.dispatch("auth/setReAuthenticationState", "not-started");
+          // if the user is still logged in, log them out
+          // if the user is not logged in, then just throw an error
+          if (store.getters["auth/isAuthenticated"]) {
+            store.dispatch("auth/autoLogoutUser");
+            throw new Error("AutoLogout! " + err.message);
```

Comment:
what about?

```
// We want to cancel all the pending axios requests. The workaround for this is to
// return a never resolving/rejecting promise so no more API calls can occur
// https://github.com/axios/axios/issues/583#issuecomment-504317347
return new Promise(() => {});
```

Reply:
We can do that as well. It doesn't really matter in this case because the user would be auto logged out.
Throwing an error will just allow it to be caught at the location where the API call is made.
For example, if the API call was made in Plio.vue
```js
await PlioAPIService.getPlio(...)
    .then(() => { ... })
    .catch( (e) => console.log(e) )
```
This will be logged here and that's it.

As the user will be logged out, no more API call would be scheduled anyway.